tests/unit/test_graph_server.py

```diff
@@ -0,0 +1,419 @@
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_outer_key_not_string():
+    result = graph_server.convert_graph_data(
+        {
+            0: {
+                "0": json.dumps({"foo": 1, "bar": 2}),
+            }
+        }
+    )
+    assert result == {"error": "Expected outer key to be str, got <class 'int'>"}
+
+
+@pytest.mark.skipif(
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_outer_value_not_dict():
+    result = graph_server.convert_graph_data({"result": 0})
+    assert result == {"error": "Expected outer value to be dict, got <class 'int'>"}
+
+
+@pytest.mark.skipif(
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_inner_key_not_string():
+    result = graph_server.convert_graph_data(
+        {
+            "result": {
+                0: json.dumps({"foo": 1, "bar": 2}),
+            }
+        }
+    )
+    assert result == {"error": "Expected inner key to be str, got <class 'int'>"}
+
+
+@pytest.mark.skipif(
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_inner_value_not_string():
+    result = graph_server.convert_graph_data(
+        {
+            "result": {
+                "0": 1,
+            }
+        }
+    )
+    assert result == {"error": "Expected inner value to be str, got <class 'int'>"}
+
+
+@pytest.mark.skipif(
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_one_column_one_row_two_columns():
+    result = graph_server.convert_graph_data(
+        {
+            "result1": {
+                "0": json.dumps(row_alex_owns_account),
+            },
+            "result2": {
+                "0": json.dumps(row_alex_owns_account),
+            },
+        }
+    )
+    assert result == {
+        "error": "Query has multiple columns - graph visualization not supported"
+    }
+
+
+@pytest.mark.skipif(
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_empty_dict():
+    result = graph_server.convert_graph_data({})
+    assert result == {
+        "error": "query result with no columns is not supported for graph visualization"
+    }
+
+
+@pytest.mark.skipif(
+    graph_visualization is None, reason="Requires `spanner-graph-notebook`"
+)
+def test_convert_wrong_row_index():
+    result0 = graph_server.convert_graph_data(
+        {
+            "result": {
+                "0": json.dumps(row_alex_owns_account),
+            }
+        }
+    )
+
+    # Changing the index should not impact the result.
+    result1 = graph_server.convert_graph_data(
+        {
+            "result": {
+                "1": json.dumps(row_alex_owns_account),
+            }
+        }
+    )
+
+    assert result1 == result0
+
+
+class TestGraphServer(unittest.TestCase):
+    def setUp(self):
+        if graph_visualization is not None:
```

Comment:
We have no coverage for when `graph_visualization` is None. If that's expected, we can add a comment to skip from coverage. See: https://github.com/googleapis/python-bigquery-magics/blob/f3cca3ab2ab47d699ae463256717343d7b84a79a/.coveragerc#L29

Reply:
```suggestion
        if graph_visualization is not None:  # pragma: NO COVER
```